datavault/backend.py

```diff
@@ -216,7 +439,378 @@ def hasMore(self, pos):
+        else:
+            struct_data = self.dataset[start:start+limit]
+        return struct_data, start + struct_data.shape[0]
+
+    def __len__(self):
+        return self.dataset.shape[0]
+
+    def hasMore(self, pos):
+        return pos < len(self)
+
+class SimpleHDF5Data(HDF5MetaData):
+    """Basic dataset backed by HDF5 file.
+
+    This is a very simple implementation that only supports a single 2-D dataset
+    of all floats.  HDF5 files support multiple types, multiple dimensions, and
+    a filesystem-like tree of datasets within one file.  Here, the single dataset
+    is stored in /DataVault within the HDF5 file.
+    """
+    def __init__(self, fh):
+        self._file = fh
+        if 'Version' not in self.file.attrs:
+            self.file.attrs['Version'] = np.asarray([2, 0, 0], dtype=np.int32)
+        self.version = np.asarray(self.file.attrs['Version'], dtype=np.int32)
+
+    def initialize_info(self, title, indep, dep):
+        ncol = len(indep) + len(dep)
+        dtype = [('f{}'.format(idx), np.float64) for idx in range(ncol)]
+        if 'DataVault' not in self.file:
+            self.file.create_dataset('DataVault', (0,), dtype=dtype, maxshape=(None,))
+        HDF5MetaData.initialize_info(self, title, indep, dep)
+
+    @property
+    def file(self):
+        return self._file()
+
+    @property
+    def dataset(self):
+        return self.file["DataVault"]
+
+    def addData(self, data):
+        """Adds one or more rows or data from a 2D array of floats."""
+        new_rows = data.shape[0]
+        old_rows = self.dataset.shape[0]
+        #if data.shape[1] != len(self.dataset.dtype):
+        #    raise errors.BadDataError(len(self.dataset.dtype), data.shape[1])
+
+        self.dataset.resize((old_rows + new_rows,))
+        #new_data = np.zeros((new_rows,), dtype=self.dataset.dtype)
+        #for col in range(data.shape[1]):
+        #    field = "f%d" % (col,)
+        #    new_data[field] = data[:,col]
+        self.dataset[old_rows:(old_rows + new_rows)] = data
+
+    def getData(self, limit, start, transpose, simpleOnly):
+        """Get up to limit rows from a dataset."""
+        if transpose:
+            raise RuntimeError("Transpose specified for simple data format: not supported")
+        if limit is None:
+            struct_data = self.dataset[start:]
+        else:
+            struct_data = self.dataset[start:start+limit]
+        columns = []
+        for idx in range(len(struct_data.dtype)):
+            columns.append(struct_data['f{}'.format(idx)])
+        data = np.column_stack(columns)
+        return data, start + data.shape[0]
+
+    def __len__(self):
+        return self.dataset.shape[0]
+
+    def hasMore(self, pos):
+        return pos < len(self)
+
+def open_hdf5_file(filename):
+    """Factory for HDF5 files.  
+
+    We check the version of the file to construct the proper class.  Currently, only two
+    options exist: version 2.0.0 -> legacy format, 3.0.0 -> extended format.
```

Comment:
Yes.  If you use the old call to datavault.new(), you will generate a
version 2.0 dataset, that only supports 2D arrays, and is API compatible
with the old functions.  When you call new_ex() you generate a version 3.0
dataset that supports all the new features but may not work with the old
add() and get() calls.  I have now made it so that if you use the new API
to create a version 3 dataset that happens to be a 2-D array of floats, it
will still work with the old API, so maybe this isn't needed any more.



On Thu, Jun 25, 2015 at 6:56 PM, Julian Kelly <notifications@github.com>
wrote:

> In datavault/backend.py
> <https://github.com/martinisgroup/servers/pull/187#discussion_r33314095>:
>
> > +        for idx in range(len(struct_data.dtype)):
> > +            columns.append(struct_data['f{}'.format(idx)])
> > +        data = np.column_stack(columns)
> > +        return data, start + data.shape[0]
> > +
> > +    def __len__(self):
> > +        return self.dataset.shape[0]
> > +
> > +    def hasMore(self, pos):
> > +        return pos < len(self)
> > +
> > +def open_hdf5_file(filename):
> > +    """Factory for HDF5 files.
> > +
> > +    We check the version of the file to construct the proper class.  Currently, only two
> > +    options exist: version 2.0.0 -> legacy format, 3.0.0 -> extended format.
>
> So version 1 is data taken before this version of the data vault
> version 2 is the old 2d array data sets stored in hdf5
> and version 3 is the fullhdf5?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/servers/pull/187/files#r33314095>.
>